src/Compiler/Utilities/Caches.fs

```diff
@@ -0,0 +1,315 @@
+// LinkedList uses nulls, so we need to disable the nullability warnings for this file.
+namespace FSharp.Compiler.Caches
+
+open System
+open System.Collections.Generic
+open System.Collections.Concurrent
+open System.Threading
+open System.Diagnostics
+open System.Diagnostics.Metrics
+
+[<Struct; RequireQualifiedAccess; NoComparison; NoEquality>]
+type CacheOptions =
+    {
+        /// Total capacity, determines the size of the underlying store.
+        TotalCapacity: int
+
+        /// Safety margin size as a percentage of TotalCapacity.
+        HeadroomPercentage: int
+    }
+
+    static member Default =
+        {
+            TotalCapacity = 128
+            HeadroomPercentage = 50
+        }
+
+[<Sealed; NoComparison; NoEquality>]
+[<DebuggerDisplay("{ToString()}")>]
+type CachedEntity<'Key, 'Value> =
+    val mutable Key: 'Key
+    val mutable Value: 'Value
+    val mutable AccessCount: int64
+    val mutable Node: LinkedListNode<CachedEntity<'Key, 'Value>>
```

Comment:
Use a Nullable type instead of passing `Unchecked.defaultof` ?

Reply:
Or even voption.